### PR TITLE
OCPBUGS-37540: Gather Azure Logs through load balancer

### PR DESCRIPTION
**Pull request by patrickdillon**
The Azure bootstrap node cannot have a public IP address, as
Azure does not allow a public IP when the VM is attached to a load
balancer with an outbound rule. Instead, CAPZ creates an inbound
nat rule to allow SSH access through the load balancer.

This PR encapsulates the logic for gathering the bootstrap
host address into its own function and adds a condition to use
the API LB hostname when the bootstrap node cannot have a public ip.

### CODE REVIEW

**Comment by patrickdillon** [comment]
Local testing was able to collect the gather bundle successfully. 

```shell
ERROR Invalid log bundle or the bootstrap machine could not be reached and bootstrap logs were not collected 
INFO Bootstrap gather logs captured here "c/log-bundle-20240807220816.tar.gz" 
```

The log bundle was incorrectly identified as invalid. Looking more closely, the file structure of the log bundle is strange. For example, one of the service files was saved as:

```
c/log-bundle-20240807220816/log-bundle-20240807220816/bootstrap/services/release-image.json
```
Obviously this looks strange, but also does not match the regex we're using to discover services: 

https://github.com/openshift/installer/blob/5804b029698ce1640deee2c9f609b44727973161/pkg/gather/service/analyze.go#L17-L22

I will be interested to see if this issue is replicated in ci or if it is a local issue for me.

**Comment by r4f4** [comment]
aws: still using the bootstrap IP address:
```
time="2024-08-07T23:26:54Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-07T23:26:54Z" level=debug msg="bootstrap manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-2d061-wdkqp-bootstrap.yaml]"
time="2024-08-07T23:26:54Z" level=debug msg="found bootstrap address: [35.92.146.200 10.0.139.79]"
time="2024-08-07T23:26:54Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-2d061-wdkqp-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-2d061-wdkqp-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-2d061-wdkqp-master-2.yaml]"
time="2024-08-07T23:26:54Z" level=debug msg="found master address: [10.0.37.150]"
time="2024-08-07T23:26:54Z" level=debug msg="found master address: [10.0.109.54]"
time="2024-08-07T23:26:54Z" level=debug msg="found master address: [10.0.24.174]"
```
vsphere: didn't use the bootstrap manifest, so it use the api dns? Maybe `BootstrapHasPublicIP` is not set?
```
time="2024-08-07T23:22:52Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-07T23:22:52Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-e3f50-4kdh4-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-e3f50-4kdh4-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-e3f50-4kdh4-master-2.yaml]"
time="2024-08-07T23:22:52Z" level=debug msg="found master address: [10.38.220.10 fd65:a1a8:60ad:1255::58 fd65:a1a8:60ad:1255:c40:1821:36d9:783b]"
time="2024-08-07T23:22:52Z" level=debug msg="found master address: [10.38.220.52 fd65:a1a8:60ad:1255::4f fd65:a1a8:60ad:1255:670c:63e1:ecb9:2052]"
time="2024-08-07T23:22:52Z" level=debug msg="found master address: [10.38.220.13 fd65:a1a8:60ad:1255::46 fd65:a1a8:60ad:1255:c7c1:b8ec:8481:f47e]"
```
azure: didn't use bootstrap manifest, used api dns instead. No complaints about invalid log bundle and the bundle looks normla.
```
time="2024-08-07T23:54:01Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-07T23:54:01Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-bdefc-vprc6-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-bdefc-vprc6-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-f5g5kq9i-bdefc-vprc6-master-2.yaml]"
time="2024-08-07T23:54:01Z" level=debug msg="found master address: [10.0.0.4]"
time="2024-08-07T23:54:01Z" level=debug msg="found master address: [10.0.0.5]"
time="2024-08-07T23:54:01Z" level=debug msg="found master address: [10.0.0.6]"
[...]
time="2024-08-07T23:54:04Z" level=info msg="Pulling debug logs from the bootstrap machine"
time="2024-08-07T23:54:04Z" level=debug msg="Added /tmp/bootstrap-ssh2650378855 to installer's internal agent"
time="2024-08-07T23:54:04Z" level=debug msg="Added /tmp/.ssh/ssh-privatekey to installer's internal agent"
[...]
```

File: cmd/openshift-install/create.go [hunk 1]
```diff
@@ -384,6 +384,7 @@ func addRouterCAToClusterCA(ctx context.Context, config *rest.Config, directory
 }
 
 func waitForBootstrapComplete(ctx context.Context, config *rest.Config) *clusterCreateError {
+	return newClientError(errors.New("BUGGIN'"))
```
Comment (jhixson74) [inline cmd/openshift-install/create.go:387]: This should probably be something else ;-0)

**Comment by r4f4** [comment]
/hold

Until @patrickdillon removes the DNM commit.

**Comment by r4f4** [comment]
openstack: didn't use bootstrap manifest, but api dns.
```
time="2024-08-08T00:26:35Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-08T00:26:35Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-f5g5kq9i-84513-jsbpc-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-f5g5kq9i-84513-jsbpc-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-f5g5kq9i-84513-jsbpc-master-2.yaml]"
time="2024-08-08T00:26:35Z" level=debug msg="found master address: [10.0.3.120]"
time="2024-08-08T00:26:35Z" level=debug msg="found master address: [10.0.1.137]"
time="2024-08-08T00:26:35Z" level=debug msg="found master address: [10.0.1.25]"
[...]
time="2024-08-08T00:28:44Z" level=info msg="Failed to gather bootstrap logs: failed to connect to the bootstrap machine: dial tcp 38.102.83.11:22: connect: connection timed out"
time="2024-08-08T00:28:54Z" level=error msg="Bootstrap failed to complete: BUGGIN'"
time="2024-08-08T00:28:54Z" level=error msg="Failed to create a kubernetes client."
time="2024-08-08T00:28:54Z" level=error msg="Invalid log bundle or the bootstrap machine could not be reached and bootstrap logs were not collected"
```

File: cmd/openshift-install/create.go (hunk 1 above)
Comment (r4f4) [reply cmd/openshift-install/create.go:387]: It was on purpose to force a bootstrap failure so we can see the bootstrap gather collection exercised in CI.

**Comment by r4f4** [comment]
@patrickdillon maybe we can always use the api dns if `publish == External` and not rely on `BootstrapHasPublicIP`?

**Comment by patrickdillon** [comment]
> openstack: didn't use bootstrap manifest, but api dns.

You were right about vsphere, BootstrapHasPublicIP is false; same is true for OpenStack. So I don't think we can use that as a flag to indicate ssh through LB... Need to rework the solution. 

> maybe we can always use the api dns if publish == External and not rely on BootstrapHasPublicIP?

I suspect most platforms don't load balance on port 22; and that's probably a good thing. We (or CAPZ) just have to do it on Azure because Azure networking is dumb and won't let us attach a public IP to a VM when it is in a LB with an outbound rule.

**Comment by r4f4** [comment]
> > openstack: didn't use bootstrap manifest, but api dns.
> 
> You were right about vsphere, BootstrapHasPublicIP is false; same is true for OpenStack. So I don't think we can use that as a flag to indicate ssh through LB... Need to rework the solution.

Either rework the solution or make sure the `BootstrapHasPublicIP` usage is consistent in all platforms. This was something we introduced late in the CAPI development, so I wouldn't be surprised that other platforms don't even know about it.

> > maybe we can always use the api dns if publish == External and not rely on BootstrapHasPublicIP?
> 
> I suspect most platforms don't load balance on port 22; and that's probably a good thing. We (or CAPZ) just have to do it on Azure because Azure networking is dumb and won't let us attach a public IP to a VM when it is in a LB with an outbound rule.

Ah right,  I missed that important detail.

**Comment by patrickdillon** [comment]
Still testing this locally but wanted to push it to give time for testing & review. 

I changed the `BootstrapHasPublicIP` function to `PublicGatherEndpoint` and from returning a bool to returning an enum. This is still consistent with how we were using BootstrapHasPublicIP but gives a better representation of the differences between platforms.

**Comment by patrickdillon** [comment]
Local testing worked. I will fix the linting issues on my next push

**Comment by r4f4** [comment]
`aws`: seems to be working as expected in terms of IP selection. Not sure why the ssh itself failed:
```
time="2024-08-08T15:20:22Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-08T15:20:22Z" level=debug msg="bootstrap manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-17vp84mv-2d061-2dxr4-bootstrap.yaml]"
time="2024-08-08T15:20:22Z" level=debug msg="found bootstrap address: [34.221.202.198 10.0.152.140]"
time="2024-08-08T15:20:22Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-17vp84mv-2d061-2dxr4-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-17vp84mv-2d061-2dxr4-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-17vp84mv-2d061-2dxr4-master-2.yaml]"
time="2024-08-08T15:20:22Z" level=debug msg="found master address: [10.0.32.25]"
time="2024-08-08T15:20:22Z" level=debug msg="found master address: [10.0.92.49]"
time="2024-08-08T15:20:22Z" level=debug msg="found master address: [10.0.59.78]"
[...]
time="2024-08-08T15:20:25Z" level=info msg="Pulling debug logs from the bootstrap machine"
time="2024-08-08T15:20:25Z" level=debug msg="Added /tmp/.ssh/ssh-privatekey to installer's internal agent"
time="2024-08-08T15:20:25Z" level=debug msg="Added /tmp/bootstrap-ssh2920318023 to installer's internal agent"
time="2024-08-08T15:20:25Z" level=info msg="Failed to gather bootstrap logs: failed to connect to the bootstrap machine: dial tcp 34.221.202.198:22: connect: connection refused"
```
`vsphere`: seems to be working as expected:
```
time="2024-08-08T15:27:39Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-08T15:27:39Z" level=debug msg="bootstrap manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-0ab48-572rm-bootstrap.yaml]"
time="2024-08-08T15:27:39Z" level=debug msg="found bootstrap address: [10.93.43.227 fd65:a1a8:60ad:1148:f3ea:7644:4517:85c0]"
time="2024-08-08T15:27:39Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-0ab48-572rm-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-0ab48-572rm-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-0ab48-572rm-master-2.yaml]"
time="2024-08-08T15:27:39Z" level=debug msg="found master address: [10.93.43.172 fd65:a1a8:60ad:1148:c11f:1911:5c39:4d4f]"
time="2024-08-08T15:27:39Z" level=debug msg="found master address: [10.93.43.221 fd65:a1a8:60ad:1148:8770:fb93:e554:e286]"
time="2024-08-08T15:27:39Z" level=debug msg="found master address: [10.93.43.186 fd65:a1a8:60ad:1148:de0a:d4ca:a300:588f]"
[...]
time="2024-08-08T15:27:39Z" level=info msg="Pulling debug logs from the bootstrap machine"
[...]
time="2024-08-08T15:27:48Z" level=debug msg="Log bundle written to /var/home/core/log-bundle-20240808152739.tar.gz"
```

**Comment by r4f4** [comment]
`nutanix`: seems to be working as expected
```
time="2024-08-08T15:47:27Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-08T15:47:27Z" level=debug msg="bootstrap manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-1cc2b-9svjv-bootstrap.yaml]"
time="2024-08-08T15:47:27Z" level=debug msg="found bootstrap address: [10.0.201.106]"
time="2024-08-08T15:47:27Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-1cc2b-9svjv-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-1cc2b-9svjv-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-ci-op-hhxwf5fi-1cc2b-9svjv-master-2.yaml]"
time="2024-08-08T15:47:27Z" level=debug msg="found master address: [10.0.202.169]"
time="2024-08-08T15:47:27Z" level=debug msg="found master address: [10.0.202.13]"
time="2024-08-08T15:47:27Z" level=debug msg="found master address: [10.0.201.240]"
[...]
time="2024-08-08T15:47:27Z" level=info msg="Pulling debug logs from the bootstrap machine"
[...]
time="2024-08-08T15:47:36Z" level=debug msg="Log bundle written to /var/home/core/log-bundle-20240808154727.tar.gz"
```

**Comment by r4f4** [comment]
`openstack`: got the IP from the bootstrap manifest. Failed as expected since the bootstrap node only has a private IP.
```
time="2024-08-08T16:55:25Z" level=debug msg="Looking for machine manifests in /tmp/installer/.clusterapi_output"
time="2024-08-08T16:55:25Z" level=debug msg="bootstrap manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-hhxwf5fi-d2a0e-r2vl9-bootstrap.yaml]"
time="2024-08-08T16:55:25Z" level=debug msg="found bootstrap address: [10.0.1.212]"
time="2024-08-08T16:55:25Z" level=debug msg="master machine manifests found: [/tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-hhxwf5fi-d2a0e-r2vl9-master-0.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-hhxwf5fi-d2a0e-r2vl9-master-1.yaml /tmp/installer/.clusterapi_output/Machine-openshift-cluster-api-guests-hhxwf5fi-d2a0e-r2vl9-master-2.yaml]"
time="2024-08-08T16:55:25Z" level=debug msg="found master address: [10.0.3.125]"
time="2024-08-08T16:55:25Z" level=debug msg="found master address: [10.0.0.91]"
time="2024-08-08T16:55:25Z" level=debug msg="found master address: [10.0.3.84]"
[...]
time="2024-08-08T16:55:25Z" level=info msg="Pulling debug logs from the bootstrap machine"
[...]
time="2024-08-08T16:57:35Z" level=info msg="Failed to gather bootstrap logs: failed to connect to the bootstrap machine: dial tcp 10.0.1.212:22: connect: connection timed out"
time="2024-08-08T16:57:36Z" level=error msg="Bootstrap failed to complete: BUGGIN'"
time="2024-08-08T16:57:36Z" level=error msg="Failed to create a kubernetes client."
time="2024-08-08T16:57:36Z" level=error msg="Invalid log bundle or the bootstrap machine could not be reached and bootstrap logs were not collected"
```

**Comment by patrickdillon** [comment]
> aws: seems to be working as expected in terms of IP selection. Not sure why the ssh itself failed:

My local aws test passed, but there are multiple AWS tests all failing in ci, so will take a closer look...

**Comment by patrickdillon** [comment]
> > aws: seems to be working as expected in terms of IP selection. Not sure why the ssh itself failed:
> 
> My local aws test passed, but there are multiple AWS tests all failing in ci, so will take a closer look...

Triple checked and the installer is trying to connect to the correct endpoint.

**Comment by patrickdillon** [comment]
> > > aws: seems to be working as expected in terms of IP selection. Not sure why the ssh itself failed:
> > 
> > 
> > My local aws test passed, but there are multiple AWS tests all failing in ci, so will take a closer look...
> 
> Triple checked and the installer is trying to connect to the correct endpoint.

Oh, I think the issue is that I'm failing the install so early in the process that the VM has not fully ignited yet and the ssh agent is not running, so this is not a problem in the code

**Comment by r4f4** [comment]
> > > aws: seems to be working as expected in terms of IP selection. Not sure why the ssh itself failed:
> > 
> > 
> > My local aws test passed, but there are multiple AWS tests all failing in ci, so will take a closer look...
> 
> Triple checked and the installer is trying to connect to the correct endpoint.

So I guess you can drop the DNM commit then and I'll LGTM.

**Comment by r4f4** [comment]
But I'll ask that we keep the hold until the feature PRs get merged first.

**Comment by patrickdillon** [comment]
+1. dropped the dnm commit. I'm fine with getting this in late tonight or tomorrow after features merge, but we do want this soon so that we can debug any failures we see in CI

File: pkg/infrastructure/clusterapi/clusterapi.go [hunk 2]
```diff
@@ -522,6 +512,30 @@ func (i *InfraProvider) ExtractHostAddresses(dir string, config *types.InstallCo
 	return nil
 }
 
+func (i *InfraProvider) getBootstrapAddress(config *types.InstallConfig, manifestsDir string) (string, error) {
+	// If the bootstrap node cannot have a public IP address, we
+	// SSH through the load balancer, as is this case on Azure.
+	if i.impl.PublicGatherEndpoint() == APILoadBalancer && config.Publish != types.InternalPublishingStrategy {
```
Comment (sadasu) [inline pkg/infrastructure/clusterapi/clusterapi.go:518]: Should the 2nd condition here be config.PublicAPI() ?

**Pull request merged**